File: 365.水壶问题.go
/*
 * @lc app=leetcode.cn id=365 lang=golang
 *
 * [365] 水壶问题
 */
package algorithmcamp

// @lc code=start
func canMeasureWater(x int, y int, z int) bool {
	for ; z != 0 && y != 0 && z < x+y; x, y, z = y, x%y, z%y {
	}
	return z == 0 || z == x+y
}

// @lc code=end

File: 365.水壶问题.rs
/*
 * @lc app=leetcode.cn id=365 lang=rust
 *
 * [365] 水壶问题
 */

// @lc code=start
impl Solution {
    pub fn can_measure_water(x: i32, y: i32, z: i32) -> bool {
        fn gcd(x: i32, y: i32) -> i32 {
            if x % y == 0 {
                return y;
            } else {
                return gcd(y, x % y);
            }
        }
        x + y >= z && !(x == 0 || y == 0) && z % gcd(x, y) == 0 || z == 0
    }
}
// @lc code=end
